Add asset manager (#745)
schultek/jaspr: docs/content/concepts/assets.mdx

---
title: Assets
description: Handling assets like images, videos, etc. in content-driven sites.
---

When building content-driven websites, you often have assets like images, videos, or other files that belong to a specific page or section. Managing these assets can become difficult if they are all stored in a central directory.

The `AssetManager` allows you to **co-locate your assets** with your content files. It handles the resolution of relative paths in your content and data, ensuring that they point to the correct location in the final build.

Additionally, it can process your assets during the build, like adding a hash for cache busting, scaling and optimizing images and more.

---

## Setup

To use the `AssetManager`, you need to set it up in your `main.dart` file.

1. Create an instance of `AssetManager`.
2. Add the `assetManager.middleware` to the server app.
3. Add `assetManager.dataLoader` and `assetManager.pageExtension` to your `ContentApp`.

```dart
import 'package:jaspr/server.dart';
import 'package:jaspr_content/jaspr_content.dart';

void main() {

>   // 1. Initialize AssetManager

Reviewer: ```suggestion
  // 1. Initialize and configure AssetManager.
```

When building custom components, you can resolve asset paths using the `context.resolveAsset()` extension method.

```dart
import 'package:jaspr/jaspr.dart';
import 'package:jaspr_content/jaspr_content.dart';

class MyComponent extends StatelessComponent {
  @override
  Iterable<Component> build(BuildContext context) sync* {
    // Resolve 'image.png' relative to the current page.
    final imagePath = context.resolveAsset('image.png');

    yield img(src: iconPath);
      
    // You can also resolve absolute paths (relative to the asset root).
    final logoPath = context.resolveAsset('/images/logo.png');
      
    yield img(src: logoPath);
  }
}
```

## Asset Transformers

Assets are processed during the build. By default, `AssetManager` uses the `HashingAssetTransformer`, which adds a content hash to the filename for cache busting (e.g., `image.a1b2c3d4.png`).

You can customize this behavior or add your own transformers by passing a list of `AssetTransformer`s to the `AssetManager` constructor.

```dart
final assetManager = AssetManager(
  directory: 'content',
  assetTransformers: [
    // Add your custom transformer
    MyCustomTransformer(),
    // Use default hashing

>     HashingAssetTransformer(),

Reviewer: ```suggestion
    // Add a custom transformer.
    MyCustomTransformer(),
    // Use the default hashing behavior.
    HashingAssetTransformer(),
```

  ],
);
```

### Custom Transformers

To create a custom transformer, extend `AssetTransformer` and implement the `transform` method.

```dart
class MyCustomTransformer extends AssetTransformer {
  @override
  Asset transform(Asset asset, [Object? aspect]) {
    // Transform the asset (e.g. compress images, optimize svg, etc.)
    // Return a new Asset or the same asset.
    return asset;
  }
}
```

> The `aspect` option is passed through as a parameter from `context.resolveAsset(aspect: ...)` and can be used to generate asset variants, e.g. differently resized images.

Reviewer: ```suggestion
The `aspect` option is passed through as a parameter from `context.resolveAsset(aspect: ...)` and can be used to generate asset variants, such as differently resized images.
```
